dbms/src/Functions/FunctionsStringSearch.h

```diff
@@ -36,61 +36,146 @@ namespace DB
+                // an escape char and it's after an escape char, we still lower it
+                // For example: "AA" (escape 'A'), -> "Aa"
+                if (char_to_lower != escape_char || escaped) {
+                    chars[i] = toLowerIfAlphaASCII(char_to_lower);
+                } else {
+                    escaped = true;
+                    continue;
+                }
+            }
+            else
+            {
+                if ((chars[i] == static_cast<unsigned char>(escape_char)) && !escaped)
+                {
+                    escaped = true;
+
+                    // It should be `chars[i] = toUpperIfAlphaASCII(chars[i])`,
+                    // but 'actual_escape_char' is always equal to 'toUpperIfAlphaASCII(str[i])'
+                    chars[i] = actual_escape_char;
+                    continue;
+                }
+                size_t utf8_len = UTF8::seqLength(char_to_lower);
+                i += utf8_len - 1;
+            }
+            escaped = false;
+        }
+    }
+
+    static void lowerColumnConst(ColumnConst * lowered_col_const, const ColumnPtr & column_escape)
     {
         auto * col_data = typeid_cast<ColumnString *>(&lowered_col_const->getDataColumn());
         RUNTIME_ASSERT(col_data != nullptr, "Invalid column type, should be ColumnString");
 
-        lowerStrings(col_data->getChars());
+        lowerColumnStringImpl(col_data, column_escape);
     }
 
-    static void lowerColumnString(MutableColumnPtr & col)
+    static void lowerColumnString(MutableColumnPtr & col, const ColumnPtr & column_escape)
     {
         auto * col_vector = typeid_cast<ColumnString *>(&*col);
         RUNTIME_ASSERT(col_vector != nullptr, "Invalid column type, should be ColumnString");
 
-        lowerStrings(col_vector->getChars());
+        lowerColumnStringImpl(col_vector, column_escape);
     }
 
-    // Only lower the 'A', 'B', 'C'...
-    static void lowerAlphaASCII(Block & block, const ColumnNumbers & arguments)
+    static void lowerColumnStringImpl(ColumnString * lowered_col_data, const ColumnPtr & column_escape)
     {
-        MutableColumnPtr column_haystack = block.getByPosition(arguments[0]).column->assumeMutable();
-        MutableColumnPtr column_needle = block.getByPosition(arguments[1]).column->assumeMutable();
-
-        auto * col_haystack_const = typeid_cast<ColumnConst *>(&*column_haystack);
-        auto * col_needle_const = typeid_cast<ColumnConst *>(&*column_needle);
+        if (column_escape == nullptr)
+        {
+            lowerStrings(lowered_col_data->getChars());
+            return;
+        }
 
-        if (col_haystack_const != nullptr)
-            lowerColumnConst(col_haystack_const);
-        else
-            lowerColumnString(column_haystack);
+        const auto * col_escape_const = typeid_cast<const ColumnConst *>(&*column_escape);
+        RUNTIME_CHECK_MSG(col_escape_const != nullptr, "escape char column should be constant");
```

Comment:
what about escape is null?

Reply:
> what about escape is null?

TiDB will raise error when escape is null.